### ISSUE TITLE
Debian: Bleachbit package leaving cruft behind after removal (i.e. 'apt purge bleachbit')

### ISSUE THREAD

**Issue by JedMeister** [bleachbit/bleachbit#1130]
When removing the (current v4.2.0) bleachbit Debian 10 package (`bleachbit_4.2.0-0_all_debian10.deb`), `/usr/share/bleachbit/__pycache__` is left behind. The specific message (full output in context below) is:

```
dpkg: warning: while removing bleachbit, directory '/usr/share/bleachbit/__pycache__' not empty so not removed
```

Obviously this isn't a huge issue, but IMO it would be nice if this dir too was removed on package removal.

**To Reproduce**

Install:
```
sudo apt install ./bleachbit_4.2.0-0_all_debian10.deb
```
Run it, then uninstall:
```
sudo apt purge bleachbit
```
```
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages will be REMOVED:
  bleachbit*
0 upgraded, 0 newly installed, 1 to remove and 1 not upgraded.
After this operation, 2,484 kB disk space will be freed.
Do you want to continue? [Y/n] y
(Reading database ... 277906 files and directories currently installed.)
Removing bleachbit (4.2.0-0) ...
dpkg: warning: while removing bleachbit, directory '/usr/share/bleachbit/__pycache__' not empty so not removed
Processing triggers for mime-support (3.62) ...
Processing triggers for gnome-menus (3.31.4-3) ...
Processing triggers for desktop-file-utils (0.23-4) ...
```

**Expected behavior**
The whole `/usr/share/bleachbit` file tree removed on package removal.

**Additional context**
(Sys info - after re-installation)
```
user@my-pc ~/Downloads$ bleachbit --sysinfo
BleachBit version 4.2.0
Git revision d929e94
GTK version 3.24.5
GTK theme = Adwaita
GTK icon theme = gnome
GTK prefer dark theme = False
SQLite version 3.27.2
local_cleaners_dir = None
locale_dir = /usr/share/locale/
options_dir = /home/user/.config/bleachbit
personal_cleaners_dir = /home/user/.config/bleachbit/cleaners
system_cleaners_dir = /usr/share/bleachbit/cleaners
locale.getdefaultlocale = ('en_AU', 'UTF-8')
os.getenv('DESKTOP_SESSION') = gnome
os.getenv('LOGNAME') = user
os.getenv('USER') = user
os.getenv('SUDO_UID') = None
os.path.expanduser('~') = /home/user
platform.linux_distribution() = ('debian', '10.8', '')
platform.platform = Linux-4.19.0-14-amd64-x86_64-with-debian-10.8
platform.version = #1 SMP Debian 4.19.171-2 (2021-01-30)
sys.argv = ['/usr/bin/bleachbit', '--sysinfo']
sys.executable = /usr/bin/python3
sys.version = 3.7.3 (default, Jul 25 2020, 13:03:44) 
[GCC 8.3.0]
__file__ = /usr/share/bleachbit/Diagnostic.py
```

**Comment by kenrap** [comment]
Umm? Isn't it up to the system package manager (such as apt) to remove any leftover cruft when uninstalling a package?

In particular, that `__pycache__` directory would be compiled bytecode leftovers from running Bleachbit in Debian?

**Comment by JedMeister** [comment]
> Umm? Isn't it up to the system package manager (such as apt) to remove any leftover cruft when uninstalling a package?

Yes it is. But it's up to the specific package maintainer to tell the package manager to do that! apt/dpkg are very careful and won't do anything that they're not explicitly told to do.

> 
> In particular, that `__pycache__` directory would be compiled bytecode leftovers from running Bleachbit in Debian?

Yes exactly! Hence why it should be removed by the package manager - via the specific packaging scripts provided in the package; bleachbit in this case.

**Comment by abitrolly** [comment]
`dh_python` is there https://github.com/bleachbit/bleachbit/blame/master/debian/debian.rules#L77 but commented out since import into `git` history. It is mentioned in the packaging policy https://www.debian.org/doc/packaging-manuals/python-policy/#dh-python but it hard to me to allocate the time to get deep into it.

**Comment by JedMeister** [comment]
Sorry, my bad. I did a poor job of sharing my context. What I actually meant was the `pybuild` tool (in the `dh-python` package - as a build depends), rather than `dh_python`. Deep apologies on any confusion I might have caused.

I'm no Debian packaging expert (actually far from it) but IIRC the cool "pybuild" automation also requires compatibility 10+.

The packages I build are fairly simple CLI python apps (for Linux only), but FYI here's the template rules file I use (COMPAT=10, build depends on `dh-python` & `python3-debian`). With a simple `setup.py` something like this:

```
[...]

setup(
    [...]
    packages=["mylib"],
    scripts=["run-me"]
)
```
And a rules file like:

```
#! /usr/bin/make -f

%:
	dh $@ --buildsystem=pybuild
```

That's it! :smile:

That will build the lib `mylib` (and put it in /usr/lib) and the executable `mun-me` (and put it in /usr/bin). If other files are needed, you can use an `install` file (one file/dir per line, `source destination` (and perhaps a `links file if I want to create any symlinks).

Clearly Bleachbit is much more complex than I deal with. And I also have the advantage that I'm only building for Debian stable.

I don't have access to my dev servers ATM, but when I'm back at my desk, I might even have a little play and see what I can see... Do you just target Debian stable, Ubuntu LTS & Ubuntu latest? Anything else? Also, out of interest, are you directly involved in the package in the Debian repos? If not, perhaps it's worth speaking with them directly? They probably know much more about Debian packaging than me. If so, then perhaps posting on the "mentors" mailing list for pointers?

To be clear though, IMO this is a really minor (bordering on insignificant IMO) issue. It's not ideal, but it's not that big a deal...

**Comment by kenrap** [comment]
> Clearly Bleachbit is much more complex than I deal with. And I also have the advantage that I'm only building for Debian stable.

The codebase is actually not that bad. It may take a week to get a feel for where everything is (granted I'm NEET atm with a lot of free time). After that you can pretty much solve any easy-to-intermediate level of Linux (or Debian) related issues that pop up. Also using `pdb` is very helpful too.

**Comment by az0** [comment]
@JedMeister 
If I provide a .deb file, would you please be able to test it again?


Here's a draft of a major overhaul to the debian.rules file. 

```
#!/usr/bin/make -f
# -*- makefile -*-

# Enable verbose mode for debugging
# export DH_VERBOSE=1

# The '%' is a pattern rule that matches any target (like 'build', 'clean', 'binary').
# This will automatically use dh_python3 to handle bytecode and __pycache__ cleanup.
%:
	dh $@ --buildsystem=pybuild

override_dh_auto_build:
	# Run the original command to remove Windows-specific files.
	$(MAKE) delete_windows_files
	# Let dh continue with the default build actions.
	dh_auto_build

override_dh_install:
	# First, run the default installation provided by debhelper.
	dh_install

	# Now, create and install the admin .desktop file.
	cp $(CURDIR)/org.bleachbit.BleachBit.desktop $(CURDIR)/debian/bleachbit/usr/share/applications/org.bleachbit.BleachBit-root.desktop
	sed -i -e 's/Name=BleachBit$$/Name=BleachBit as Administrator/g' $(CURDIR)/debian/bleachbit/usr/share/applications/org.bleachbit.BleachBit-root.desktop
	sed -i -e 's/^Exec=bleachbit$$/Exec=pkexec bleachbit/g' $(CURDIR)/debian/bleachbit/usr/share/applications/org.bleachbit.BleachBit-root.desktop
```

Also, the update involves adding `debhelper-compat (= 13)` in the `Build-Depends:` section of  the `debian/control` file.
